Add --seed(-s) parameter for reproducible problem generation

simonusher opened the pull request:
Hello!
We needed to have reproducible problem generation in our work so I added this parameter. 
If you feel it would be useful to have something like this then please review this.

I don't really have much experience with Rust programming so there could be mistakes. I also might have removed more than you wanted (codomain.rs line 131) to make the compiler happy. I pass a mutable rng there, so I couldn't use the parallel progress iterator.

simonusher committed:
add rng seed

simonusher committed:
refactor structopt seed, add get_rng to remove code duplication

simonusher committed:
remove code duplication related to get_rng from codomain.rs, move get_rng to configuration.rs

tobiasvandriessel commented on src/problem/codomain.rs:
```diff
@@ -72,31 +73,40 @@ enum CodomainCommand {
-            for folder_path in folder_paths {
-                handle_folder(folder_path)?;
+    match codomain_opt {
+        CodomainOpt { codomain_command, seed } => {
```
You can probably move the multiple `let mut rng = get_rng(seed);` lines here, just after the `seed` is defined.

tobiasvandriessel replied:
Also, I'd not do a match if there's no other options, although it _is_ more clear what properties the CodomainOpt struct has. So, I'd do it much like I did it before, except for a line `let mut rng = get_rng(codomain_opt.seed)` inserted right after the function signature declaration.

tobiasvandriessel commented on src/problem/problem_generation.rs:
```diff
@@ -90,48 +94,63 @@ enum ProblemCommand {
-                generate_codomain_and_problem_from_folder(
-                    &folder_path,
+    match problem_opt {
+        ProblemOpt { problem_command, seed } => {
```
Again, you can probably move the multiple let mut rng = get_rng(seed); lines here, just after the seed is defined.

tobiasvandriessel replied:
Again, I'd not do a match if there's no other options, although it is more clear what properties the CodomainOpt struct has. So, I'd do it much like I did it before, except for a line let mut rng = get_rng(codomain_opt.seed) inserted right after the function signature declaration.

tobiasvandriessel commented on src/problem/codomain.rs:
```diff
@@ -118,8 +128,8 @@ fn handle_folder(folder_path: PathBuf) -> Result<(), Box<dyn Error>> {
     //And handle each of them
-    file_entries.into_par_iter().progress().for_each(|path| {
-        handle_input_configuration_file(path).unwrap();
+    file_entries.into_iter().for_each(|path| {
```
It's a shame the parallel iterator needs to go, but it's probably worth it given the added feature of reproducible randomness. I do wonder whether the `for_each` will always process the files in the same order and whether that matters then for the reproducibility.

tobiasvandriessel replied:
And didn't the progress bar work anymore? I think I'd like to keep that actually.

tobiasvandriessel reviewed (changes requested):
Thanks for creating a pull request for this; I actually spoke about this feature with somebody that's working on integrating this into a testing framework, so thank you for taking the time to make this work already :)

I do have some small comments, but nothing major. Would you be able to process my review comments? I don't mind doing it myself, but perhaps it's nice when you do it, since it's your pull request. Happy to help!

simonusher committed:
remove unnecessary matches, remove duplicated get_rng, restore progress bar in codomain generation, add explicit sort for file_entries for reproducibility

simonusher committed:
add sorted() calls after each read_dir to ensure reproducibility

simonusher commented:
Thanks for your reply. I addressed your comments, removed the unnecessary match statements, and moved the get_rng call up. I also restored the progress bar. I must have removed it by accident.
Next, taking what you said about file reading order into consideration, I checked the [documentation](https://doc.rust-lang.org/std/fs/fn.read_dir.html). It states that the order in which entries are returned from read_dir is not guaranteed,  so I added an explicit sorted() call, in places where you didn't sort the entries already. I'm convinced that it shouldn't break anything and ensure reproducible results, but I would really appreciate your review here. 
Lastly, the binary size also got a bit bigger, probably due to more imports. I hope that's not a problem.

Commit:
use .sorted() everywhere we currently use sort()

tobiasvandriessel commented:
Thanks @simonusher, I appreciate the effort you put in. Apologies for my late reply.
Really nice that you sorted the directory entries, that's very convenient! I changed my usage of sort() at some places to use .sorted() as well. Now I'll be testing whether the changes indeed do not break anything; for this I'll need to add some integration tests, so this might take me a bit of time.

Commit:
add integration test

tobiasvandriessel commented:
@simonusher I've added a simple integration test, but I should definitely add more (unit & integration) tests later on. I consider the PR closed as the integration test works on 2 PCs and my LTGA experiment runner (private project) seems to be able to work nicely still, so the file IO is working as it should.

tobiasvandriessel commented:
FYI: You can test whether the integration test is successful by running 'cargo test'.

simonusher commented:
I'm sorry for not replying earlier. Thank you for incorporating this into your library. I hope it will find its uses.

tobiasvandriessel commented:
No problem, and I am sure it will :)
